### DISCUSSION TITLE
그래프

### DISCUSSION THREAD

**Discussion by vumrra** [subak-ba/cs-study#23] [자료 구조]
# 그래프

그래프 알고리즘은 경로찾기, 최적화 문제 등 다양한 문제를 해결하는데 사용되는 중요한 알고리즘 입니다.

> 알고리즘: 문제 해결 방법을 정의한 일련의 단계적 절차이자 문제를 해결하기 위한 동작들의 모임

## 그래프란?

그래프는 정점과 간선으로 구성된 자료구조 입니다.

방향성이 있는 방향그래프와 방향성이 없는 무방향 그래프로 나뉩니다.

추가로 간선에는 가중치가 있을 수 있습니다.

![image](https://github.com/subak-ba/cs-study/assets/131235625/f3226338-7a49-4572-ba6b-ccb23bbad101)

## 그래프 용어 정리

정점(노드)

- 그래프에서 하나의 점을 나타냅니다.

간선

- 그래프에서 정점과 정점을 연결하는 선을 나타냅니다.

인접

- 두개의 정점이 간선으로 직접 연결되어있는 상테를 나타냅니다.

차수

- 해당 노드에 연결된 간선의 수를 나타냅니다.

경로

- 그래프에서 정점들을 연결하는 간선의 순서를 나타냅니다. 경로의 길이는 경로에 속하는 간선의 수입니다.

사이클

- 그래프에서 동일한 노드로 되돌아오는 경로를 말합니다.
- 경로의 시작 노드와 끝 노드가 동일한 경우를 말합니다.

가중치

- 가중치 그래프에서 간선에 할당된 값을 나타냅니다.

## 그래프 구현 방법

그래프를 구현하는 방법은 인접 리스트, 인접 행렬을 사용하여 구현할 수 있습니다.

### 인접 리스트

인접리스트는 각 정점에 대해 연결된 모든 정점의 리스트를 저장하는 방식입니다.

특정 정점과 인접한 정점들을 한눈에 파악할 수 있다는 장점이 있습니다.

### 인접 행렬

인접 행렬은 2차원 배열을 사용하여 노드 간의 연결 관계를 행렬 형태로 나타내는 방식입니다.

노드간의 연결중인지 여부를 빠르게 확인 가능하지만 모든 정점에 연결 관계를 기록하기 때문에 특히 정점의 수가 많고 간선의 수가 적을때 메모리 소비가 크다는 단점이 있습니다.

인접 행렬은 간선이 많은 그래프에서 주로 사용하고 인접 리스트는 상대적으로 간선이 적은 그래프에서 주로 사용합니다.

![image](https://github.com/subak-ba/cs-study/assets/131235625/7f49ff50-3dd1-42d5-9a52-2ad015752bdb)

### 그래프 탐색 알고리즘

그래프에서 특정 정점을 찾는 알고리즘 입니다. 그래프 내의 특정한 정점을 찾기 위해서 해당 그래프의 정점을 순회합니다.

기본적인 탐색 알고리즘은 **너비 우선 탐색**과 **깊이 우선 탐색**이 있습니다.

### 최단 경로 알고리즘

가중치가 있는 그래프에서 두 정점간의 최소 가중치 합을 가지는 경로를 찾는 알고리즘입니다.

대표적인 최단 경로 알고리즘으로는 다익스트라 알고리즘, 벨만-포드 알고리즘, 플로이드 알고리즘등이 있습니다.

최단 경로 알고리즘은 네트워크 설계나 교통 시스템 최적화, 네비게이션 시스템 등 다양한 분야에서 경로 최적화 문제를 해결하는데 활용됩니다.

> *다익스트라 알고리즘: 가중치 그래프에서 시작 정점으로 부터 모든 장점까지의 최단 경로를 찾는 알고리즘
> 

> *벨만-포드 알고리즘: 다익스트라 알고리즘와 동일하지만 음수 가중치를 가진 간선이 있는 경우에도 동작합| 니다.
> 

> *크루스칼 알고리즘: 가중치 그래프에서 최소 신장을 갖는 트리를 찾는 알고리즘입니다. 모든 정점을 가중치 순으로 정렬한 뒤, 사이클을 형성하지 않는 간선을 추가하여 트리를 형성합니다.
> 

> *프림 알고리즘: 가중치 그래프에서 최소 신장을 갖는 트리를 찾는 알고리즘입니다. 시작 정점에서 출발하여 현재 트리와 연결되지 않은 노드 중 최소 가중치의 간선을 선택하여 트리를 확장합니다.
> 

그래프 탐색 알고리즘 중에 BFS, DFS를 자바 코드로 설명해보겠습니다.

### 깊이 우선 탐색 Depth First Search

1. 정점 방문
2. 방문한 정점의 인접 정점 중 아직 방문하지 않은 정점을 찾아 DFS를 재귀적으로 수행한다.

### 넓이 우선 탐색 Breadth First Search

1. 장점 방문시 해당 정점을 큐에 삽입
2. 삽입된 정점의 이웃 중 아직 방문하지 않은 정점을 차례대로 방문
3. 2번 작업이 끝났다면 큐에 있는 정점을 꺼내서 2번을 다시 실행

## DFS

![image](https://github.com/subak-ba/cs-study/assets/131235625/cdb81cf6-3b59-462b-b858-de97e49dc951)

탐색시 한 경로의 끝까지 탐색 후 탐색할 정점이 없다면 다시 가장 가까운 갈림길로 돌아와서(백 트랙킹) 아직 방문하지 않은 방향으로 다시 탐색을 진행합니다.

예시를 미로로 든다면 미로를 탐색할 때 해당 길에서 갈 수 있을때까지 계속 가다가 길이 막힌다면 다시 가장 가까운 갈림길로 돌아와서 다른 방향으로 출구를 찾는 방식입니다.

DFS는 스택 자료구조를 사용하며 모든 노드를 방문하려할 때 사용합니다.

BFS보다 간단하게 구현 가능하지만 검색 속도는 느립니다.

> 스택: 후입선출 LIFO로 데이터를 보관하는 자료구조 입니다.
> 

```java
import java.util.*;
import java.lang.*;
import java.io.*;

import java.util.*;

class Graph {
    private int V; // 정점의 수
    private LinkedList<Integer> adj[]; // 인접 리스트

    Graph(int v) {
        V = v;
        adj = new LinkedList[v]; // 입력받은 정점의 수로 인접 리스트 초기화
        for (int i = 0; i < v; ++i)
            adj[i] = new LinkedList<>();
    }

    // 간선 추가 함수
    void addEdge(int v, int w) {
        adj[v].add(w);
    }

    // DFS 구현
    void DFS(int v) {
        boolean visited[] = new boolean[V];
        DFSUtil(v, visited);
    }

    // DFS 탐색 함수 (재귀)
    private void DFSUtil(int v, boolean visited[]) {
        visited[v] = true;
        System.out.print(v + " ");

        for (int n : adj[v]) {
            if (!visited[n]) {
                DFSUtil(n, visited);
            }
        }
    }
}

public class Main {
    public static void main(String args[]) {
        Graph g = new Graph(7); // 정점의 개수: 7

        g.addEdge(0, 1);
        g.addEdge(0, 2);
        g.addEdge(1, 3);
        g.addEdge(1, 4);
        g.addEdge(2, 5);
        g.addEdge(2, 6);

        g.DFS(0); // 시작 정점를 0으로 설정
    }
}

```

- 시작 정점을 입력 받습니다
- 해당 정점과 인접 한 경로의 모든 정점을 탐색합니다 (재귀)
- 탐색할 정점이 없다면 돌아가서(백 트래킹) 방문하지 않은 정점부터 탐색을 시작하여 탐색을 이어나갑니다.

![image](https://github.com/subak-ba/cs-study/assets/131235625/31a27233-24c8-465d-aa04-8847619e6e1b)

BFS는 재귀적으로 동작하지 않고 큐 자료구조를 사용하여 방문한 정점들을 차례대로 저장한 후 꺼내 사용합니다

> 큐: 선입선출 FIFO으로 데이터를 보관하는 자료구조입니다.

시작 정점에서 가장 가까운 정점을 먼저 방문한 후 멀리 있는 정점은 나중에 방문하는 방식이기 때문에 탐색하려는 노드 사이의 최단 경로를 찾을 때 주로 사용합니다.

```java
import java.util.*;

class Graph {
    private int V; // 정점의 수
    private LinkedList<Integer> adj[]; // 인접 리스트

    // 생성자
    Graph(int v) {
        V = v;
        adj = new LinkedList[v];
        for (int i = 0; i < v; ++i)
            adj[i] = new LinkedList<>();
    }

    // 간선 추가 함수
    void addEdge(int v, int w) {
        adj[v].add(w);
    }

    // BFS 구현
    void BFS(int s) {
        boolean visited[] = new boolean[V];
        LinkedList<Integer> queue = new LinkedList<>();

        visited[s] = true;
        queue.add(s);

        while (!queue.isEmpty()) {
            s = queue.poll();
            System.out.print(s + " ");

            for (int n : adj[s]) {
                if (!visited[n]) {
                    visited[n] = true;
                    queue.add(n);
                }
            }
        }
    }
}

public class Main {
    public static void main(String args[]) {
        Graph g = new Graph(7); // 정점의 개수: 7

        g.addEdge(0, 1);
        g.addEdge(0, 2);
        g.addEdge(1, 3);
        g.addEdge(1, 4);
        g.addEdge(2, 5);
        g.addEdge(2, 6);

        g.BFS(0); // 시작 정점을 0으로 설정
    }
}

```

- 시작 정점을 큐에 저장합니다
- 해당 정점을 꺼내서 해당 정점의 인접 정점들을 방문하여 큐에 저장합니다
- 모두 방문 했다면 큐에서 꺼내서 꺼낸 정점에서 인접 정점을 탐색하는 방식으로 탐색을 이어나갑니다

참고자료

https://binco.tistory.com/entry/Java-%EC%95%8C%EA%B3%A0%EB%A6%AC%EC%A6%98-DFS%EC%99%80BFS-%EC%99%84%EB%B2%BD%EC%A0%95%EB%A6%AC)](https://binco.tistory.com/entry/Java-%EC%95%8C%EA%B3%A0%EB%A6%AC%EC%A6%98-DFS%EC%99%80BFS-%EC%99%84%EB%B2%BD%EC%A0%95%EB%A6%AC

https://velog.io/@boyeon_jeong/%EA%B7%B8%EB%9E%98%ED%94%84-%EC%A2%85%EB%A5%98-%EB%B0%8F-%EA%B0%9C%EB%85%90)](https://velog.io/@boyeon_jeong/%EA%B7%B8%EB%9E%98%ED%94%84-%EC%A2%85%EB%A5%98-%EB%B0%8F-%EA%B0%9C%EB%85%90

**Comment by onee-only** [comment]
👍

**Reply by onee-only** [reply]
예제 추가 해 주실 수 있나요?
